base/dict.jl

```diff
@@ -99,6 +99,10 @@ mutable struct Dict{K,V} <: AbstractDict{K,V}
 end
 function Dict{K,V}(kv) where V where K
     h = Dict{K,V}()
+    chklen = IteratorSize(kv)
+    if (chklen isa HasLength) || (chklen isa HasShape)
```

Comment:
Are both `HasLength` and `HasShape` necessary here? Wouldn't length alone be sufficient?

Reply:
No, because `HasLength` is not a subtype of `HasShape`, but morally all `HasShape` implement `length`.

Reply:
I have to confess, I didn't follow the Cambrian explosion of collections traits, but it seems kind of bizarre that checking for `HasLength` is not sufficient to know if you can call `length` on something. But carry on 🤷‍♂️

Reply:
Btw I stole this directly from https://github.com/JuliaLang/julia/blob/453a7ddcb6894cfcd8fc016634857b39416585cc/base/asyncmap.jl#L127-L128

Reply:
I've seen `Base.haslength(iter)` used as well: https://github.com/JuliaLang/julia/blob/a9be54d1b089182a6b079224e9cab58a318b9541/base/generator.jl#L98

Reply:
Thanks @garborg! I wasn't aware of `haslength`